File: journey_log/LinkedList/AddTwoNumbers02.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result_head = new ListNode();   // dummy head node
        ListNode result = result_head;
        int carry = 0;
        
        while (l1 != null || l2 != null) {
            int sum = 0;
            
            if (l1 != null) {
                sum += l1.val;
                l1 = l1.next;
            }
            
            if (l2 != null) {
                sum += l2.val;
                l2 = l2.next;
            }
            
            sum += carry;
            
            if (sum > 9) {
                carry = 1;
                sum = sum % 10;
            } else {
                carry = 0;
            }
            
            ListNode curr = new ListNode(sum);
            
            result.next = curr;
            result = result.next;
        }
        
        if (carry == 1) {
            ListNode msd = new ListNode(carry);
            result.next = msd;
        }
        
        return result_head.next;
    }
}

File: journey_log/LinkedList/AddTwoNumbers02.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* result = new ListNode();
        ListNode* node = result;
        int carry = 0;
        
        while (l1 != nullptr || l2 != nullptr) {
            int l1_val = 0;
            int l2_val = 0;
            int sum;
            
            if (l1 != nullptr) {
                l1_val = l1->val;
                l1 = l1->next;
            }
            
            if (l2 != nullptr) {
                l2_val = l2->val;
                l2 = l2->next;
            }
            
            sum = carry + l1_val + l2_val;
            
            if (sum > 9) {
                carry = 1;
                sum = sum % 10;
            } else {
                carry = 0;
            }
            
            ListNode* curr = new ListNode(sum);
            node->next = curr;
            node = node->next;
        }
        
        if (carry == 1) {
            ListNode* msb = new ListNode(carry);
            node->next = msb;
        }
        
    return result->next;
    }
};
